operator/controller/dnsnameresolver-crd/controller.go

```diff
@@ -0,0 +1,132 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "dnsnameresolver_crd_controller"
+)
+
+var (
+	controllerLog = ctrl.Log.WithName(controllerName)
+)
+
+// New creates and returns a controller that creates DNSNameResolver.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		cache:  mgr.GetCache(),
+		client: mgr.GetClient(),
+		config: config,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	clusterNamePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		crd := manifests.DNSNameResolverCRD().DeepCopy()
+		expectedName := types.NamespacedName{Name: crd.Name}
+		actualName := types.NamespacedName{
+			Name: o.GetName(),
+		}
+		return expectedName == actualName
+	})
+	if err := c.Watch(source.Kind(mgr.GetCache(), &apiextensionsv1.CustomResourceDefinition{}),
+		&handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DependentCaches is a list of caches that are used by Controllers watching DNSNameResolver
+	// resources. The dnsnameresolver_crd controller starts these caches once
+	// the DNSNameResolver CRD has been created.
+	DependentCaches []cache.Cache
+	// DependentControllers is a list of controllers that watch DNSNameResolver
+	// resources. The dnsnameresolver_crd controller starts these controllers once
+	// the DNSNameResolver CRD has been created and the DependentCaches are started.
+	DependentControllers []controller.Controller
+}
+
+// reconciler handles the actual CRD reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+
+	cache            cache.Cache
+	client           client.Client
+	startCaches      sync.Once
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a CRD and creates or
+// reconciles the DNSNameResolver CRD.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	controllerLog.Info("reconciling", "request", request)
+
+	// Ensure the DNSNameResolver CRD is created.
+	if err := r.ensureDNSNameResolverCRD(ctx); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Start the dependent caches and wait for the caches to sync.
+	r.startCaches.Do(func() {
+		var wg sync.WaitGroup
+		for i := range r.config.DependentCaches {
+			cache := &r.config.DependentCaches[i]
+
+			// Start the dependent cache.
+			go func() {
+				if err := (*cache).Start(ctx); err != nil {
+					controllerLog.Error(err, "cannot start cache")
```

Comment:
I will change the code so that if the caches fail to start or sync then the controller won't be started. However, as we are doing this only once across all reconciles, this process won't be retried.

Reply:
As discussed in [this thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on this one in a separate PR.